### PR TITLE
[Feature] Omnisource training

### PR DESCRIPTION
**Pull request by kennymckormick**

### CODE REVIEW

File: mmaction/core/runner/omnisource_runner.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            if isinstance(mode, str):  # self.train()
+                if not hasattr(self, mode):
+                    raise ValueError(
+                        f'runner has no method named "{mode}" to run an '
+                        'epoch')
+                epoch_runner = getattr(self, mode)
+            else:
+                raise TypeError(
+                    'mode in workflow must be a str, but got {}'.format(
+                        type(mode)))
```
Comment (dreamerlin) [inline mmaction/core/runner/omnisource_runner.py:153]: f-string

File: mmaction/datasets/base.py [hunk 2]
```diff
@@ -37,6 +38,14 @@ class BaseDataset(Dataset, metaclass=ABCMeta):
             from 0. Default: 1.
         modality (str): Modality of data. Support 'RGB', 'Flow'.
             Default: 'RGB'.
+        sample_by_class (bool): Sampling by class, should be set `True` when
+            you want to perform inter-class data balancing. Only compatible
```
Comment (dreamerlin) [inline mmaction/datasets/base.py:42]: `you want to perform` -> `performing`

File: mmaction/datasets/base.py [hunk 3]
```diff
@@ -37,6 +38,14 @@ class BaseDataset(Dataset, metaclass=ABCMeta):
...
+            with `multi_class == False`. Only applies for training. Default:
+            False.
+        power (float): We support sampling data with the probability
+            proportional to the power of its label frequency (freq ^ power).
```
Comment (dreamerlin) [inline mmaction/datasets/base.py:46]: `probability proportional to the power of its label frequency (freq ^ power) when sampling data.`

File: mmaction/datasets/base.py [hunk 4]
```diff
@@ -88,6 +105,15 @@ def load_json_annotations(self):
...
         return video_infos
 
+    def parse_by_class(self):
+        video_infos_by_class = {}
+        for item in self.video_infos:
+            label = item['label']
+            if label not in video_infos_by_class:
+                video_infos_by_class[label] = []
+            video_infos_by_class[label].append(item)
+        return video_infos_by_class
```
Comment (dreamerlin) [inline mmaction/datasets/base.py:115]:
```python
def parse_by_class(self):
    video_infos_by_class = defaultdict(list)
    for item in self.video_infos:
        label = item['label']
        video_infos_by_class[label].append(item)
    return video_infos_by_class 
```

File: mmaction/datasets/rawframe_dataset.py [hunk 5]
```diff
@@ -74,6 +75,14 @@ class RawframeDataset(BaseDataset):
         num_classes (int): Number of classes in the dataset. Default: None.
         modality (str): Modality of data. Support 'RGB', 'Flow'.
                             Default: 'RGB'.
+        sample_by_class (bool): Sampling by class, should be set `True` when
+            you want to perform inter-class data balancing. Only compatible
+            with `multi_class == False`. Only applies for training. Default:
+            False.
+        power (float): We support sampling data with the probability
```
Comment (dreamerlin) [inline mmaction/datasets/rawframe_dataset.py:82]: (float | None)

File: mmaction/datasets/rawvideo_dataset.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+
+    Args:
+        ann_file (str): Path to the annotation file.
+        pipeline (list[dict | callable]): A sequence of data transforms.
+        start_index (int): Specify a start index for frames in consideration of
+            different filename format. However, when taking videos as input,
+            it should be set to 0, since frames loaded from videos count
+            from 0. Default: 0.
+        **kwargs: Keyword arguments for ``BaseDataset``.
```
Comment (dreamerlin) [inline mmaction/datasets/rawvideo_dataset.py:51]: `clipname_tmpl` is missing and `start_index` is redundant.

File: mmaction/apis/train.py [hunk 7]
```diff
@@ -33,19 +36,48 @@ def train_model(model,
...
-        build_dataloader(ds, **dataloader_setting) for ds in dataset
-    ]
+
+    if cfg.omnisource:
+        # The option can override videos_per_gpu
+        train_ratio = cfg.data.get('train_ratio', None)
+        omni_videos_per_gpu = cfg.data.get('omni_videos_per_gpu', None)
+        dataloader_setting_tmpl = dict(
+            videos_per_gpu=cfg.data.get('videos_per_gpu', {}),
+            workers_per_gpu=cfg.data.get('workers_per_gpu', {}),
```
Comment (dreamerlin) [inline mmaction/apis/train.py:46]: These two lines are not correct. (`{}` can not be set as default value for these two)

File: mmaction/core/runner/omnisource_runner.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+def cycle(iterable):
+    iterator = iter(iterable)
+    while True:
+        try:
+            yield next(iterator)
+        except StopIteration:
+            iterator = iter(iterable)
+
+
```
Comment (dreamerlin) [inline mmaction/core/runner/omnisource_runner.py:18]: `@HOOKS.register_module()`

File: mmaction/core/runner/omnisource_runner.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    def before_epoch(self, runner):
+        for data_loader in runner.data_loaders:
+            if hasattr(data_loader.sampler, 'set_epoch'):
+                # in case the data loader uses `SequentialSampler` in Pytorch
+                data_loader.sampler.set_epoch(runner.epoch)
+            elif hasattr(data_loader.batch_sampler.sampler, 'set_epoch'):
+                # batch sampler in pytorch wraps the sampler as its attributes.
+                data_loader.batch_sampler.sampler.set_epoch(runner.epoch)
+
+
```
Comment (dreamerlin) [inline mmaction/core/runner/omnisource_runner.py:30]: @RUNNERS.register_module()

File: mmaction/core/runner/omnisource_runner.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+            self._iter += 1
+
+        self.call_hook('after_train_epoch')
+        self._epoch += 1
+
+    # Now that we use validate hook, not implement this func to save efforts.
+    def val(self, data_loader, **kwargs):
+        raise NotImplementedError
+
+    def run(self, data_loaders, workflow, max_epochs=None, **kwargs):
```
Comment (dreamerlin) [inline mmaction/core/runner/omnisource_runner.py:107]: docstring for `max_epochs` is missing.

File: mmaction/datasets/pipelines/loading.py [hunk 11]
```diff
@@ -1115,7 +1160,30 @@ def __call__(self, results):
 
         results['audios'] = np.array(resampled_clips)
         results['audios_shape'] = results['audios'].shape
+        return results
+
+
+@PIPELINES.register_module()
+class PseudoClipBuilder(object):
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:1167]: `(object)` will be removed in the future for reduce redundancy.

File: mmaction/datasets/pipelines/loading.py [hunk 12]
```diff
@@ -965,6 +965,51 @@ def __call__(self, results):
 
 
 @PIPELINES.register_module()
+class ImageDecode(object):
```
Comment (dreamerlin) [inline mmaction/datasets/pipelines/loading.py:968]: ~`(object)`~

File: mmaction/core/runner/omnisource_runner.py (hunk 8 above)
Comment (kennymckormick) [reply mmaction/core/runner/omnisource_runner.py:18]: This class doesn't need registration, since we directly use it without calling any build_xxx function.

File: mmaction/core/runner/omnisource_runner.py (hunk 9 above)
Comment (kennymckormick) [reply mmaction/core/runner/omnisource_runner.py:30]: This class doesn't need registration, since we directly use it without calling any build_xxx function.

File: mmaction/core/runner/omnisource_runner.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+                    self.call_hook('before_train_iter')
+                    self.run_iter(
+                        data_batch, train_mode=True, source=f'/aux{idx}')
+                    self.call_hook('after_train_iter')
+            self._iter += 1
+
+        self.call_hook('after_train_epoch')
+        self._epoch += 1
+
+    # Now that we use validate hook, not implement this func to save efforts.
```
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/core/runner/omnisource_runner.py:103]: However, it was not used in any config of action recognition models. If users want to check the validation accuracy or loss, they are recommended to use the `validate` hook rather than `val pipeline`.

File: mmaction/apis/train.py [hunk 14]
```diff
@@ -105,8 +149,25 @@ def train_model(model,
...
 
+    # All hooks are registered now, for omnisource experiments, we need to
+    # override the `end_of_epoch` method for each hook.
+    # if cfg.omnisource:
+    #
+    #     def new_end_of_epoch(runner):
+    #         return runner.inner_iter + 1 == len(runner.main_loader)
+    #
+    #     for hook in runner._hooks:
+    #         hook.end_of_epoch = new_end_of_epoch
```
[the remark this answers is not in the document]
Comment (kennymckormick) [reply mmaction/apis/train.py:160]: No, this comment is out-of-date (just to remind me to implement something ... ). I will remove them.

File: mmaction/apis/train.py [hunk 15]
```diff
@@ -63,12 +95,21 @@ def train_model(model,
...
+            work_dir=cfg.work_dir,
+            logger=logger,
+            meta=meta)
+    else:
+        runner = EpochBasedRunner(
+            model,
+            optimizer=optimizer,
+            work_dir=cfg.work_dir,
+            logger=logger,
+            meta=meta)
```
Comment (dreamerlin) [inline mmaction/apis/train.py:112]:
```python
Runner = OmniSourceRunner if cfg.onmisource else EpochBasedRunner
runner = Runner(...)
```

File: mmaction/apis/train.py [hunk 16]
```diff
@@ -33,19 +36,48 @@ def train_model(model,
...
+            ]
+
+    else:
+        dataloader_setting = dict(
+            videos_per_gpu=cfg.data.get('videos_per_gpu', 2),
+            workers_per_gpu=cfg.data.get('workers_per_gpu', 0),
+            # cfg.gpus will be ignored if distributed
+            num_gpus=len(cfg.gpu_ids),
+            dist=distributed,
+            seed=cfg.seed)
```
Comment (dreamerlin) [inline mmaction/apis/train.py:74]: The `dataloader_setting_tmpl` can be merged.

File: mmaction/apis/train.py [hunk 17]
```diff
@@ -109,4 +153,11 @@ def train_model(model,
...
         runner.load_checkpoint(cfg.load_from)
-    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
+    if cfg.omnisource:
+        runner.run(
+            data_loaders,
+            cfg.workflow,
+            cfg.total_epochs,
+            train_ratio=train_ratio)
+    else:
+        runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
```
Comment (dreamerlin) [inline mmaction/apis/train.py:163]: can we use a `runner_kwargs` dict to merge them ?

Comment (kennymckormick) [reply mmaction/apis/train.py:163]: That's OK

**Comment by kennymckormick** [comment]
@innerlee , would u please help review this PR, thx a lot!

**Comment by innerlee** [comment]
will take a look at Friday

File: mmaction/apis/train.py [hunk 18]
```diff
@@ -33,19 +36,37 @@ def train_model(model,
...
 
-    data_loaders = [
-        build_dataloader(ds, **dataloader_setting) for ds in dataset
-    ]
+    if cfg.omnisource:
+        # The option can override videos_per_gpu
+        train_ratio = cfg.data.get('train_ratio', None)
+        omni_videos_per_gpu = cfg.data.get('omni_videos_per_gpu', None)
+        if omni_videos_per_gpu is None:
+            dataloader_settings = [dataloader_setting] * len(dataset)
```
Comment (innerlee) [inline mmaction/apis/train.py:54]: `dataloader_setting`s in the `dataloader_settings` list are all referenced to a same object.
pls double check that this is intended

File: mmaction/datasets/base.py [hunk 19]
```diff
@@ -105,7 +130,12 @@ def dump_results(self, results, out):
 
     def prepare_train_frames(self, idx):
         """Prepare the frames for training given the index."""
-        results = copy.deepcopy(self.video_infos[idx])
+        if self.sample_by_class:
+            # Then, the idx is the class index
+            samples = self.video_infos_by_class[idx]
+            results = copy.deepcopy(random.choice(samples))
```
Comment (innerlee) [inline mmaction/datasets/base.py:136]: use np.random for consistency

File: mmaction/datasets/pipelines/loading.py [hunk 20]
```diff
@@ -964,6 +964,52 @@ def __call__(self, results):
...
+    """
+
+    def __init__(self, io_backend='disk', decoding_backend='cv2', **kwargs):
+        self.io_backend = io_backend
+        self.decoding_backend = decoding_backend
+        self.kwargs = kwargs
+        self.file_client = None
+
+    def __call__(self, results):
+        """Perform the ``RawFrameDecode`` to pick frames given indices.
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:988]: change name in docstring

File: mmaction/datasets/pipelines/loading.py [hunk 21]
```diff
@@ -1115,7 +1161,31 @@ def __call__(self, results):
 
         results['audios'] = np.array(resampled_clips)
         results['audios_shape'] = results['audios'].shape
+        return results
+
+
+@PIPELINES.register_module()
+class PseudoClipBuilder:
```
Comment (innerlee) [inline mmaction/datasets/pipelines/loading.py:1168]: name proposal: `BuildPseudoClips`

File: mmaction/datasets/rawframe_dataset.py [hunk 22]
```diff
@@ -1,5 +1,6 @@
 import copy
 import os.path as osp
+import random
```
Comment (innerlee) [inline mmaction/datasets/rawframe_dataset.py:3]: use np.random

**Comment by innerlee** [comment]
some minor comments

File: mmaction/apis/train.py (hunk 18 above)
Comment (kennymckormick) [reply mmaction/apis/train.py:54]: Yes, that would be OK. Since dataloader settings are the same for every dataset, and bulid_dataloader will not alter it.

File: mmaction/datasets/pipelines/loading.py (hunk 21 above)
Comment (kennymckormick) [reply mmaction/datasets/pipelines/loading.py:1168]: ok, name changed to BuildPseudoClip

**Comment by innerlee** [comment]
so easy to get conflicts

**Comment by kennymckormick** [comment]
> so easy to get conflicts

resolved

**Review by innerlee — APPROVED** [review]
